frameworks/Java/armeria/src/main/java/hello/services/PostgresDbService.java

```diff
@@ -0,0 +1,161 @@
+package hello.services;
+
+import hello.models.World;
+import hello.helpers.PostgresDbHelper;
+import hello.helpers.HttpHeadersHelper;
+
+import java.util.concurrent.ThreadLocalRandom;
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.MediaType;
+import com.linecorp.armeria.server.annotation.Default;
+import com.linecorp.armeria.server.annotation.Get;
+import com.linecorp.armeria.server.annotation.Param;
+import com.linecorp.armeria.server.annotation.ProducesJson;
+
+public class PostgresDbService {
+  private static final ObjectMapper MAPPER = new ObjectMapper();
+
+  private static final String SELECT_QUERY = "SELECT * FROM world WHERE id = ?";
+  private static final String UPDATE_QUERY =
+      "UPDATE world SET randomNumber = ? WHERE id = ?";
+
+  @Get("/db")
+  @ProducesJson
+  public HttpResponse db() throws Exception {
+    return HttpResponse.of(
+        HttpHeadersHelper.getHttpHeader(MediaType.JSON_UTF_8),
+        HttpData.of(MAPPER.writeValueAsBytes(getWorld(getRandomNumber()))));
+  }
+
+  // need to use regex as /queries/{count} doesn't work when count is null
+  @Get("regex:^/queries/(?<count>.*)$")
+  @ProducesJson
+  public HttpResponse queries(
+      @Param("count")
+      @Default("")
+          String count) throws JsonProcessingException, SQLException {
+    return HttpResponse.of(
+        HttpHeadersHelper.getHttpHeader(MediaType.JSON_UTF_8),
+        HttpData.of(
+            MAPPER.writeValueAsBytes(getWorlds(getSanitizedCount(count)))));
+  }
+
+  @Get("regex:^/updates/(?<count>.*)$")
+  @ProducesJson
+  public HttpResponse update(
+      @Param("count")
+      @Default("")
+          String count) throws JsonProcessingException, SQLException {
+    return HttpResponse.of(
+        HttpHeadersHelper.getHttpHeader(MediaType.JSON_UTF_8),
+        HttpData.of(
+            MAPPER.writeValueAsBytes(
+                getUpdatedWorlds(getSanitizedCount(count)))));
+  }
+
+  private static int getRandomNumber() {
+    return 1 + ThreadLocalRandom.current().nextInt(10000);
+  }
+
+  private int getSanitizedCount(String count) {
+    try {
+      int intCount = Integer.parseInt(count);
+      if (intCount < 1) {
+        return 1;
+      }
+      if (intCount > 500) {
+        return 500;
+      }
+      return intCount;
+    } catch (NumberFormatException e) {
+      return 1;
+    }
+  }
+
+  private World getWorld(int number) throws SQLException {
+    Connection connection = DriverManager.getConnection(
```

Comment:
Oh I didn't notice this code before.  There are a couple of issues here that should be addressed.

1. **Close all of the JDBC objects you create when running a query.**

Specifically I mean the `Connection`, `PreparedStatement`, and `ResultSet`.  Use the "try with resources" language feature for this.  A block of code using JDBC tends to look like this:

```java
try (Connection connection = ...;
     PreparedStatement statement = connection.prepareStatement(...)) {
  statement.setInt(1, ...);
  try (ResultSet resultSet = statement.executeQuery()) {
    // resultSet.next() or while (resultSet.next()) ...
  }
}
```

(If you don't need to set any parameter values on the `PreparedStatement`, then the `ResultSet` declaration can be lumped into the first `try` statement.  This should be true for the fortunes query.)

Otherwise you probably have a resource leak.  I'd guess that when you benchmark this code (not in verify mode), performance deteriorates very quickly as it piles up an unbounded number of connections to the database.

2. **Use connection pooling.**

I don't know what `DriverManager.getConnection(url, username, password)` does because I've never used it, but I'm assuming it creates a *new* connection to the database.  That's an expensive thing to be doing on each request.  You want to reuse connections.  This holds true for pretty much any application that talks to a database; it's not a micro-optimization that we only make in benchmark scenarios.

I take it that the Ameria web framework doesn't come with its own JDBC connection pool.  In that case, try [HikariCP](https://brettwooldridge.github.io/HikariCP/).  Setting up the connection pool is easy:

```java
HikariConfig config = new HikariConfig();
config.setJdbcUrl("jdbc:postgresql://tfb-database:5432/hello_world");
config.setUsername("benchmarkdbuser");
config.setPassword("benchmarkdbpass");
config.setMaximumPoolSize(48); // This number is questionable but good enough
DataSource dataSource = new HikariDataSource(config);
```

A `DataSource` has a parameterless `getConnection()` method that retrieves a connection from the pool.  When you close that connection, it returns the `Connection` object to the pool without closing the underlying physical connection.

If you make this `DataSource` a non-static member of the `PostgresDbService` initialized in the constructor, then wherever you need a connection you'd get it like this:

```java
try (Connection connection = dataSource.getConnection();
     PreparedStatement statement = connection.prepareStatement(...)) {
  ...
}
```

Reply:
Thanks for the info, I should have another commit up soon.